src/ApparentHorizons/Strahlkorper.cpp

```diff
@@ -0,0 +1,399 @@
+                               const size_t l_max, const size_t m_max,
+                               std::array<double, 3> center) noexcept
+    : l_max_(l_max),
+      m_max_(m_max),
+      ylm_(l_max, m_max),
+      // clang-tidy: do not std::move trivially constructable types
+      center_(center),  // NOLINT
+      strahlkorper_coefs_(ylm_.phys_to_spec(radius_at_collocation_points)) {
+  ASSERT(radius_at_collocation_points.size() == ylm_.physical_size(),
+         "Bad size " << radius_at_collocation_points.size() << ", expected "
+                     << ylm_.physical_size());
+}
+
+template <typename Fr>
+void Strahlkorper<Fr>::pup(PUP::er& p) {
+  p | l_max_;
+  p | m_max_;
+  p | center_;
+  p | strahlkorper_coefs_;
+  // All variables other than the above are cached quantities that
+  // can be recomputed from the above.
+
+  if (p.isUnpacking()) {
+    ylm_ = YlmSpherepack(l_max_, m_max_);
+  }
+}
+
+template <typename Fr>
+std::array<double, 3> Strahlkorper<Fr>::physical_center() const noexcept {
+  // For better accuracy, one should do an integral of
+  // x^i * AreaElement divided by Area.  The method here is an approximation
+  // using only the l=1 modes.
+  // See Eq. (37) of Hemberger et al, arXiv:1211.6079 for the form of the
+  // exact physical center that we approximate here.
+  std::array<double, 3> result = center_;
+  SpherepackIterator it(l_max_, m_max_);
+  result[0] += strahlkorper_coefs_[it.set(1, 1)()] * sqrt(0.75);
+  result[1] -= strahlkorper_coefs_[it.set(1, -1)()] * sqrt(0.75);
+  result[2] += strahlkorper_coefs_[it.set(1, 0)()] * sqrt(0.375);
+  return result;
+}
+
+template <typename Fr>
+double Strahlkorper<Fr>::average_radius() const noexcept {
+  return ylm_.average(coefficients());
+}
+
+template <typename Fr>
+double Strahlkorper<Fr>::radius(const double theta, const double phi) const
+    noexcept {
+  return ylm_.interpolate_from_coefs(strahlkorper_coefs_, {{{theta, phi}}})[0];
+}
+
+template <typename Fr>
+bool Strahlkorper<Fr>::point_is_contained(const std::array<double, 3>& x) const
+    noexcept {
+  // The point is assumed to be in Cartesian coords in the Strahlkorper frame.
+  // Make the point relative to the center of the Strahlkorper.
+  auto xmc = x;
+  for (size_t d = 0; d < 3; ++d) {
+    gsl::at(xmc, d) -= gsl::at(center_, d);
+  }
+
+  // Convert the point from Cartesian to spherical coordinates.
+  const double r = sqrt(square(xmc[0]) + square(xmc[1]) + square(xmc[2]));
+  const double theta = acos(xmc[2] / r);
+  double phi = atan2(xmc[1], xmc[0]);
+  // Range of atan2 is [-pi,pi],
+  // So adjust to get range [0,2pi]
+  if (phi < 0.0) {
+    phi += 2.0 * M_PI;
+  }
+
+  // Is the point inside the surface?
+  return r < radius(theta, phi);
+}
+
+template <typename Fr>
+void Strahlkorper<Fr>::initialize_jac_and_hess() const noexcept {
+  if (jac_.get(0, 0).size() > 0) {
+    return;
+  }
+  initialize_mesh_quantities();
+
+  const auto theta_phi = ylm_.theta_phi_points();
```

Comment:
const ref?

Reply:
ylm_.theta_phi_points() doesn't return a ref, so I don't need a ref there.  If I put one there, then I think its ok, because it's a const ref to a temporary, which I believe c++ has a special rule to guarantee that it stays in scope.

Reply:
Not sure what I should do...

Reply:
Ah, fair enough.